.github/workflows/python-app.yml

```diff
@@ -105,15 +115,87 @@ jobs:
       run: |
         make lint-type-check
 
-    - name: Run Tests Coverage
+    - name: Run test on Node ${{ matrix.ci_node_index }}
       run: |
-        make coverage-report
+        set -e
+        mkdir -p test-results
+        rm -r test-results/parts || echo "no files to delete"
+        OUT=$(python -m tests --filepath test-results/results.xml --total-nodes ${{ matrix.ci_node_total }} --node-id ${{ matrix.ci_node_index }})
+        IFS=',' read -a FILE_INFO <<< "$OUT"
+        FILES=("${FILE_INFO[@]:1}")
+        echo "Selected Files are: " ${FILES[@]}
+        echo "Previous timings are: ${FILE_INFO[0]}"
+        ./coverage/coverage.sh results-${{ matrix.python }}-${{ matrix.ci_node_index }}.xml ${FILES[@]}
+
+    - name: Upload Artifact(results)
+      uses: actions/upload-artifact@v3
+      with:
+        name: results-${{ matrix.python }}-${{ matrix.ci_node_index }}
+        path: test-results/results-${{ matrix.python }}-${{ matrix.ci_node_index }}.xml
+
+    - name: Remove local test-results folder  # We don't want to create cache YET.
+      run: |
+        rm -r test-results
+
+    - name: Upload Artifact(coverage reports)
+      uses: actions/upload-artifact@v3
+      with:
+        name: coverage-${{ matrix.python }}-${{ matrix.ci_node_index }}
+        path: .coverage
 
     - name: Git Check (tests may not alter any git visible files)
       run: |
         git status -sb --porcelain
         [ -z "$(git status --porcelain)" ]
 
+  coverage:
+    needs: tests
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python: ['3.7', '3.8', '3.9']
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Set up Python ${{ matrix.python }}
+      uses: actions/setup-python@v1
+      with:
+        python-version: ${{ matrix.python }}
+
+    - name: Cache python modules
+      uses: actions/cache@v3
+      with:
+        path: ${{ env.pythonLocation }}
+        key: ${{ env.pythonLocation }}-${{ hashFiles('requirements.txt') }}-${{ hashFiles('requirements.lint.txt') }}
+
+    - name: Install dependencies
+      run: |
+        make install
+        mkdir -p test-results/parts
+
+    - name: Download all results
+      # Downloads results-*.xml,
+      uses: actions/download-artifact@v2
+    - name: Merge results
+      run: |
+        mv results-${{ matrix.python }}-*/results-${{ matrix.python }}-* test-results/parts/
+        python3 -c "from tests.utils import merge_results; merge_results('./test-results', 'results.xml')"
```

Comment:
use the `tests` module -- use sub commands to distinguish the task -- no raw python code in github actions